### DISCUSSION TITLE
Why does RHF behave differently in MUI TextField when defaultValues are present and when they are not?

### DISCUSSION THREAD

**Discussion by 37108** [react-hook-form/react-hook-form#5271] [Q&A] [answered]
I made a simple Code Sandbox example.

> https://codesandbox.io/s/polished-rain-f3lqk?file=/src/correct.tsx

### environments
- material-ui/core: 4.11.4
- react: 17.0.2
- react-dom: 17.0.2
- react-hook-form: 7.6.2

## question

When I wrote the code without using defaultValues, it works correctly, input current value is displayed on console.

```typescript
export const Greeting: VFC = () => {
  const { register, handleSubmit } = useForm();
  return (
      <form onSubmit={handleSubmit((data) => { console.log(data) })}>
        <TextField label="last_name" id="last_name" {...register("last_name")} />
        <Button variant="outlined" type="submit"> submit </Button>
      </form>
    </div>
  );
};
```

But when I wrote with defaultValues, it didn't work correctly.  
default value does not set on the input element, input value does not overwrite default and default value is displayed on console. 

```typescript
export const Greeting: VFC = () => {
  const { register, handleSubmit } = useForm({
    defaultValues: { first_name: "noah" }
  });
  return (
      <form onSubmit={handleSubmit((data) => { console.log(data) })}>
        <TextField label="last_name" id="last_name" {...register("last_name")} />
        <Button variant="outlined" type="submit"> submit </Button>
      </form>
    </div>
  );
};
```

In other words, i expect about,
- use TextField as an uncontrolled component
- the default value is shown up at input field when it's rendered
- when submit, the values are same as displayed values

So, what is the different between define without defaultValues and define with it ?
I think this is not React Hook Forms problem, and this is because of forwarding ref. but i cannot understand well...

and what is the recommended way to combine UI library components ?

## other
I know how to write it in the control component, as described in the documentation and it works as I expect.
> https://react-hook-form.com/get-started#IntegratingControlledInputs

I know if `register().ref` is passed `inputRef`, it works as I expect. what I want to say is below code. 
This is not as simple as it sounds if there are a lot of fields, so i want write more simple and use uncontrolled components.

```
export const Greeting: VFC = () => {
  const { register, handleSubmit } = useForm({
    defaultValues: { last_name: "williams" }
  });
  const { ref, ...rest } = register("last_name");

  return (
    <form onSubmit={handleSubmit((data) => {console.log(data)})}>
      <TextField label="last_name" id="last_name" inputRef={ref} {...rest} />
      <Button variant="outlined" type="submit">submit</Button>
    </form>
  );
};
```

**Comment by leapful** [comment, marked as answer]
@37108 

When using "register" from react-hook-form, the "ref" callback of "register" should be set correctly to the ref prop handling of the target component.

In your example, the TextField is something like:

```
<TextField {...register('last_name')} />

equivalent to

const { ref } = register('last_name');

<TextField ref={ref} />
```

which is incorrect because prop "ref" here is a "div" element for styling input, not the input. Without default value (undefined), the "onChange" handling will use "value" from the "onChange" event as fallback handling and works correctly but with default value, "ref" element (it's "div" element now) will be used and causes wrong value.

If you want to use uncontrolled component with "register", you need to set the correct "ref" handling , in this case, it's "inputRef" of Material-UI which is designed to be compatible with 3rd party library, otherwise, just use controlled component with "Controller" or "useController".

**Reply by 37108** [reply]
@leapful 
> which is incorrect because prop "ref" here is a "div" element for styling input, not the input. Without default value (undefined), the "onChange" handling will use "value" from the "onChange" event as fallback handling and works correctly but with default value, "ref" element (it's "div" element now) will be used and causes wrong value.

i'm so happy to understand what is happened. thanks you explained me !
